server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1383,19 +1384,20 @@ final boolean tryRenewSyncCommit() {
     @Override
     public boolean shouldPeriodicallyFlush() {
         ensureOpen();
+        final long translogGenerationOfCurrentCommit = translog.uncommittedGeneration();
+        final long uncommittedTranslogSize = translog.sizeInBytesByMinGen(translogGenerationOfCurrentCommit);
         final long flushThreshold = config().getIndexSettings().getFlushThresholdSize().getBytes();
-        final long uncommittedSizeOfCurrentCommit = translog.uncommittedSizeInBytes();
-        if (uncommittedSizeOfCurrentCommit < flushThreshold) {
+        if (uncommittedTranslogSize < flushThreshold) {
             return false;
         }
         /*
-         * We should only flush ony if the shouldFlush condition can become false after flushing.
-         * This condition will change if the `uncommittedSize` of the new commit is smaller than
-         * the `uncommittedSize` of the current commit. This method is to maintain translog only,
-         * thus the IndexWriter#hasUncommittedChanges condition is not considered.
+         * We should only flush ony if the shouldPeriodicallyFlush condition can become false after flushing.
+         * This condition will change if the new commit points to the later translog generation than the current commit's.
+         * This method is to maintain translog only, thus the IndexWriter#hasUncommittedChanges condition is not considered.
          */
-        final long uncommittedSizeOfNewCommit = translog.sizeOfGensAboveSeqNoInBytes(localCheckpointTracker.getCheckpoint() + 1);
-        return uncommittedSizeOfNewCommit < uncommittedSizeOfCurrentCommit;
+        final long translogGenerationOfNewCommit =
+            translog.getMinGenerationForSeqNo(localCheckpointTracker.getCheckpoint() + 1, false).translogFileGeneration;
```

Comment:
I'm not happy with the extra boolean flag to include / exclude the current generation as a fall back. It's too subtle an error prone. How about doing the following (I think you had it in the passed and we moved away from it towards the uncommittedX api - sorry for that):
1) If the min gen for the local checkpoint + 1 is > current committed gen , return true.
2) If the min gen is equal to the *current* translog gen, the current gen is not empty  (using `totalOperationsByMinGen`) and the local checkpoint is equal to the max seq#, return true.